pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -101,48 +102,113 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, upd
+	var services []*model.Service
+	service := push.ServiceForHostname(proxy, host.Name(key.Name))
+	// push.ServiceForHostname will return nil if the proxy doesn't care about the service OR it was deleted.
+	// we can cross-reference with WatchedResources to figure out which services were deleted.
+	if service == nil {
+		// We assume a service was deleted and delete all clusters for that service.
+		deletedClusters = append(deletedClusters, serviceClusters[key.Name].UnsortedList()...)
+		deletedClusters = append(deletedClusters, subsetClusters[key.Name].UnsortedList()...)
+	} else {
+		// Service exists. If the service update has port change, we need to the corresponding port clusters.
+		services = append(services, service)
+		for port, cluster := range servicePortClusters[service.Hostname.String()] {
+			// if this service port is removed, we can conclude that it is a removed cluster.
+			if _, exists := service.Ports.GetByPort(port); !exists {
 				deletedClusters = append(deletedClusters, cluster)
 			}
-		} else {
-			services = append(services, service)
-			// If servicePorts has this service, that means it is old service.
-			if servicePorts[service.Hostname.String()] != nil {
-				oldPorts := servicePorts[service.Hostname.String()]
-				for port, cluster := range oldPorts {
-					// if this service port is removed, we can conclude that it is a removed cluster.
-					if _, exists := service.Ports.GetByPort(port); !exists {
-						deletedClusters = append(deletedClusters, cluster)
-					}
-				}
-			}
 		}
 	}
-	clusters, log := configgen.buildClusters(proxy, updates, services)
-	return clusters, deletedClusters, log, true
+	return services, deletedClusters
+}
+
+// deltaFromDestinationRules computes the delta clusters from the updated destination rules.
+func (configgen *ConfigGeneratorImpl) deltaFromDestinationRules(updatedDr model.ConfigKey, proxy *model.Proxy, push *model.PushContext,
+	subsetClusters map[string]sets.String,
+) ([]*model.Service, []string) {
+	var deletedClusters []string
+	var services []*model.Service
+	cfg := push.DestinationRuleByName(proxy, updatedDr.Name, updatedDr.Namespace)
+	if cfg == nil {
+		// Destinationrule was deleted. Find matching services from previous destinationrule.
+		prevCfg := push.PrevDestinationRuleByName(proxy, updatedDr.Name, updatedDr.Namespace)
+		// PreviousConfig can be nil if the destination rule hosts were updated with non matching hosts.
+		if prevCfg == nil {
+			log.Debugf("Prev DestinationRule form PrevSidecarScope is missing for %s/%s", updatedDr.Namespace, updatedDr.Name)
+			return nil, nil
+		}
+		dr := prevCfg.Spec.(*networking.DestinationRule)
+		services = append(services, push.ServicesForHostname(proxy, host.Name(dr.Host))...)
+	} else {
+		// Destinationrule was updated. Find matching services from updated destinationrule.
+		dr := cfg.Spec.(*networking.DestinationRule)
+		services = append(services, push.ServicesForHostname(proxy, host.Name(dr.Host))...)
```

Comment:
This service may not be complete, think about this case a subset hostaname can be different from Dr.host

Reply:
Can you explain more? I thought even subsets are routed by same host name as DR?